doc/decisions/0_drafts/buffer_struct.md

```diff
@@ -0,0 +1,189 @@
+# Buffer Struct
+
+## Problem
+
+For this decision a buffer is a fixed-length sequence of arbitrary bytes.
+In other words, it is a pointer to data combined with the size of this data.
+
+Many parts of Elektra use such buffers, but there is currently no uniform API for all of these.
+
+## Constraints
+
+## Assumptions
+
+## Considered Alternatives
+
+### Public Buffer struct
+
+Introduce a new struct that is part of the public API and therefore declared in a public header.
+The struct should be usable everywhere, so it should be declared by `libelektra-core`.
+
+```c
+typedef struct {
+    void * data;
+    size_t size;
+} ElektraBuffer;
+```
+
+> **Note**: It is not clear yet, whether `void * data` or `const void * data` is preferrable.
+
+This struct can be embedded into other structs, e.g.:
+
+```c
+struct _Key
+{
+	ElektraBuffer name;
+    ElektraBuffer value;
+
+    KeySet * meta;
+	keyflag_t flags;
+	uint16_t refs;
+};
+```
+
+The public API for accessing the data could remain unchanged:
+
+```c
+const char * keyName (ElektraKey * key) {
+    return key->name.data;
+}
+
+const char * keyNameSize (ElektraKey * key) {
+    return key->name.size;
+}
+```
+
+It could also return the buffer in a single function.
+Either by value:
+
+```c
+const ElektraBuffer keyName (ElektraKey * key) {
+    return key->name;
+}
+
+void foo(ElektraKey * key) {
+    const ElektraBuffer name = keyName (key);
+    printf("size %zd, first part: %s", name.size, (const char*)name.data);
+}
+```
+
+or as a pointer:
+
+```c
+const ElektraBuffer * keyName (ElektraKey * key) {
+    return &key->name;
+}
+
+void foo(ElektraKey * key) {
+    const ElektraBuffer * name = keyName (key);
+    printf("size %zd, first part: %s", name->size, (const char*)name->data);
+}
+```
+
+A [short investigation](https://godbolt.org/z/r8q7YfqaW) into the generated assembly code indicates that the pointer version is likely superior.
+
+The buffer struct could also be used as an argument in the public API.
+For simple functions that would only take one such `ElektraBuffer` argument this there is little benefit.
+Especially on the caller side, the code becomes less readable:
+
+```c
+void fooNoBuffer(void * data, size_t size);
+void fooWithBuffer(ElektraBuffer buffer);
+void fooWithBufferPtr(ElektraBuffer * buffer);
+
+void caller(void) {
+    size_t size1 = 42;
+    void * data1 = malloc(size1);
+
+    fooNoBuffer(data1, size1);
+    fooWithBuffer((ElektraBuffer){.data = data1, .size = size1});
+    fooWithBufferPtr(&(ElektraBuffer){.data = data1, .size = size1});
+
+    ElektraBuffer buffer; // buffer from somewhere else
+    fooNoBuffer(buffer.data, buffer.size);
+    fooWithBuffer(buffer);
+    fooWithBufferPtr(&buffer);
```

Comment:
Not sure about this. We have to copy the buffer internally or else the data is corrupt when `buffer` goes out of scope.

Reply:
Totally depends on what `fooWithBufferPtr` does. Take this for example:

```c
void fooWithBufferPtr(ElektraBuffer * buffer) {
    buffer->size = 0;
}
```

Not sure when this could be used, but it is totally possible to have a "correct" `fooWithBufferPtr`.

Of course a `keySetName` with a buffer would need to make a copy. But `keySetName` already makes a copy, so that's nothing new.

The example is intentionally vague, because the main point was showing that `fooNoBuffer` is clearer.